jlm/llvm/opt/alias-analyses/RegionAwareMemoryNodeProvider.cpp

```diff
@@ -675,159 +603,288 @@ RegionAwareMemoryNodeProvider::Create(
+        const auto calleeCallGraphNode = callGraphNodeIndex[&lambdaNode];
+
+        // Add the edge caller -> callee to the call graph
+        callGraphSuccessors[callerIndex].Insert(calleeCallGraphNode);
+      }
+      else if (
+          dynamic_cast<const PointsToGraph::ExternalMemoryNode *>(&callee)
+          || dynamic_cast<const PointsToGraph::ImportNode *>(&callee))
+      {
+        // Add the edge caller -> node representing external functions
+        callGraphSuccessors[callerIndex].Insert(externalNodeIndex);
+      }
+    }
+  };
+
+  // Recursive function finding call nodes, adding edges to the call graph
+  const auto HandleCalls = [&](rvsdg::Region & region, size_t callerIndex, auto recurse) -> void
+  {
+    for (auto & node : region.Nodes())
+    {
+      if (is<CallOperation>(&node))
+      {
+        HandleCall(node, callerIndex);
+      }
+      else if (auto structural = dynamic_cast<rvsdg::StructuralNode *>(&node))
+      {
+        for (size_t i = 0; i < structural->nsubregions(); i++)
+        {
+          recurse(*structural->subregion(i), callerIndex, recurse);
+        }
+      }
+    }
+  };
+
+  // For all functions, visit all their calls and add outgoing edges in the call graph
+  for (size_t i = 0; i < lambdaNodes.size(); i++)
+  {
+    HandleCalls(*lambdaNodes[i]->subregion(), i, HandleCalls);
+
+    // If the function has escaped, add an edge from the node representing all external functions
+    const auto & lambdaMemoryNode = pointsToGraph.GetLambdaNode(*lambdaNodes[i]);
+    if (pointsToGraph.GetEscapedMemoryNodes().Contains(&lambdaMemoryNode))
+    {
+      callGraphSuccessors[externalNodeIndex].Insert(i);
+    }
+  }
+
+  // Used by the implementation of Tarjan's SCC algorithm
+  const auto GetSuccessor = [&](size_t nodeIndex)
+  {
+    return callGraphSuccessors[nodeIndex].Items();
+  };
+
+  // Find SCCs in the call graph
+  std::vector<size_t> sccIndex;
+  std::vector<size_t> topologicalOrder;
+  auto numSCCs = util::FindStronglyConnectedComponents<size_t>(
+      numCallGraphNodes,
+      GetSuccessor,
+      sccIndex,
+      topologicalOrder);
+
+  // sccIndex are distributed in a reverse topological order, so the sccIndex is used
+  // when creating the list of SCCs and the functions they contain
+  FunctionSCCs_.resize(numSCCs);
+  for (size_t i = 0; i < lambdaNodes.size(); i++)
+  {
+    FunctionSCCs_[sccIndex[i]].Insert(lambdaNodes[i]);
+    FunctionToSccIndex_[lambdaNodes[i]] = sccIndex[i];
+  }
+
+  // Also note which SCC contains all external functions
+  ExternalNodeSccIndex_ = sccIndex[externalNodeIndex];
+}
+
+std::vector<const rvsdg::LambdaNode *>
+RegionAwareMemoryNodeProvider::CollectLambdaNodes(const rvsdg::RvsdgModule & rvsdgModule)
+{
+  std::vector<const rvsdg::LambdaNode *> result;
+
+  // Recursively traverses all structural nodes, but does not enter into lambdas
+  const auto CollectLambdasInRegion = [&](rvsdg::Region & region, auto recurse) -> void
```

Comment:
You can avoid handing in `recurse`. Just explicitly provide the type for the lambda expression (without auto) and you should be able to recursively call it.

Reply:
I can do that, yes. There is more overhead from doing that (see [here](https://quick-bench.com/q/f49eDKZxXpISkSs6vk6vywOCAww)), but I highly doubt that runtime makes much of a difference in the grand scheme of things.

Reply:
I do not think that it makes a difference i practice. We have bigger fishes to fry at other places.